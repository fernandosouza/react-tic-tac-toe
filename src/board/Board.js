import React, { PureComponent } from 'react';
import styled from 'styled-components';
import { Slot } from './Slot';

const BoardWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  outline: var(--grid-border);
  width: var(--board-size);
`;

/* *
 * Class that renders the game board.
 * */
class Board extends PureComponent {

  /**
   * Creates a Board.
   */
  constructor(props) {
    super(props);
    this.slots = Array(9).fill(0, 0, 9);
    this.slotClick_ = this.handleSlotClick_.bind(this);
  }

  handleSlotClick_(index) {
    this.props.onSlotClick(index);
  }

  /**
   * Renders board slots that will be fufilled by gamers' pieces.
   * @private 
   */
  renderSlots_() {
    let { filledSlots, winnerSlots } = this.props;
    return this.slots.map((_, index) => {
      return (
        <Slot
          key={index}
          winner={winnerSlots.includes(index)}
          player={filledSlots.get(index)}
          index={index}
          onSlotClick={this.slotClick_}
        />
      );
    });
  }

  /**
   * @inheritdoc
   */
  render() {
    return (
      <BoardWrapper>
        {this.renderSlots_()}
      </BoardWrapper>
    );
  }
}

export default Board;
