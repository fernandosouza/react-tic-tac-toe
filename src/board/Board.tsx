import React, { PureComponent } from 'react';
import styled from 'styled-components';
import { Slot } from './Slot';

const BoardWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  border: 1px solid var(--grid-border-color);
  width: calc(var(--board-size) + 2px);
  box-shadow: -6px 26px 142px black;
  border-radius: 9px;
  overflow: hidden;
`;


/* *
 * Class that renders the game board.
 * */
class Board extends PureComponent {
  private slots = Array(9).fill(0, 0, 9);

  /**
   * Renders board slots that will be fufilled by gamers' pieces.
   * @private 
   */
  renderSlots_() {
    return this.slots.map((_, index) => {
      return (
        <Slot
          key={index}
          index={index}
        />
      );
    });
  }

  /**
   * @inheritdoc
   */
  render() {
    return (
      <BoardWrapper>
        {this.renderSlots_()}
      </BoardWrapper>
    );
  }
}

export default Board;
