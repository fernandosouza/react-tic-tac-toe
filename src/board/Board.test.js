import React from 'react';
import Board from './Board';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('Board', () => {
  let filledSlots;
  let winnerSlots;

  beforeEach(() => {
    filledSlots = new Map();
    winnerSlots = [];
  });

  it('renders 9 slots', () => {
    const board = renderer.create(
      <Board
        filledSlots={filledSlots}
        winnerSlots={winnerSlots}
      />
    ).toJSON();
    expect(board).toMatchSnapshot();
  });

  it('renders filled slots', () => {
    filledSlots.set(0, 1);
    filledSlots.set(1, 0);
    filledSlots.set(2, 2);

    const board = renderer.create(
      <Board
        filledSlots={filledSlots}
        winnerSlots={winnerSlots}
      />
    ).toJSON();
    expect(board).toMatchSnapshot();
  });

  it('should call a callback function after clicking on slots', () => {
    const callbackMock = jest.fn();
    const wrapper = mount(<Board
      filledSlots={filledSlots}
      winnerSlots={winnerSlots}
      onSlotClick={callbackMock}
    />);
    wrapper.find('.board-slot').at(0).simulate('click');
    expect(callbackMock.mock.calls.length).toBe(1);
  });

  it('should call a callback function with the slot index as argument after clicking on slots', () => {
    const callbackMock = jest.fn();
    const wrapper = mount(<Board
      filledSlots={filledSlots}
      winnerSlots={winnerSlots}
      onSlotClick={callbackMock}
    />);
    wrapper.find('.board-slot').at(1).simulate('click');
    expect(callbackMock.mock.calls[0][0]).toBe(1);
  });
});
