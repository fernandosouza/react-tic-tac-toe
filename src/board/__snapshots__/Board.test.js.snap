// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders 9 slots 1`] = `
<div
  className="board"
>
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
</div>
`;

exports[`Board renders filled slots 1`] = `
<div
  className="board"
>
  <div
    className="board-slot filled player1"
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot filled player2"
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
  <div
    className="board-slot "
    onClick={[Function]}
  />
</div>
`;
